l1-contracts/contracts/zksync/facets/Executor.sol

```diff
@@ -165,30 +161,60 @@ contract ExecutorFacet is Base, IExecutor {
                 require(logSender == L2_BOOTLOADER_ADDRESS, "bu");
                 require(_expectedSystemContractUpgradeTxHash == logValue, "ut");
-            } else {
+            } else if (logKey > uint256(SystemLogKey.EXPECTED_SYSTEM_CONTRACT_UPGRADE_TX_HASH_KEY)) {
                 revert("ul");
             }
         }
 
-        // We only require 9 logs to be checked, the 10th is if we are expecting a protocol upgrade
-        // Without the protocol upgrade we expect 9 logs: 2^9 - 1 = 511
-        // With the protocol upgrade we expect 8 logs: 2^10 - 1 = 1023
+        // Without the protocol upgrade we expect 13 logs: 2^13 - 1 = 8191
+        // With the protocol upgrade we expect 14 logs: 2^14 - 1 = 16383
         if (_expectedSystemContractUpgradeTxHash == bytes32(0)) {
-            require(processedLogs == 511, "b7");
+            require(processedLogs == 8191, "b7");
         } else {
-            require(processedLogs == 1023, "b8");
+            require(processedLogs == 16383, "b8");
         }
     }
 
+    /// @dev Check that L2 blob logs are proper and batch contain all information for them
+    /// @dev The logs processed here should line up such that only one log for each key from the
+    ///      SystemLogKey enum in Constants.sol is processed per new batch.
+    function _processBlobLogs(CommitBatchInfo calldata _newBatch) internal pure returns (bytes32[] memory blobHashes) {
+        // Copy L2 to L1 logs into memory.
+        bytes memory emittedL2Logs = _newBatch.systemLogs;
+
+        // Used as bitmap to set/check log processing happens exactly once.
+        // See SystemLogKey enum in Constants.sol for ordering.
+        uint256 processedLogs;
+
+        blobHashes = new bytes32[](MAX_NUMBER_OF_BLOBS);
+
+        // linear traversal of the logs
+        for (uint256 i = 0; i < emittedL2Logs.length; i = i.uncheckedAdd(L2_TO_L1_LOG_SERIALIZE_SIZE)) {
+            // Extract the values to be compared to/used such as the log sender, key, and value
+            (address logSender, ) = UnsafeBytes.readAddress(emittedL2Logs, i + L2_LOG_ADDRESS_OFFSET);
+            (uint256 logKey, ) = UnsafeBytes.readUint256(emittedL2Logs, i + L2_LOG_KEY_OFFSET);
+            (bytes32 logValue, ) = UnsafeBytes.readBytes32(emittedL2Logs, i + L2_LOG_VALUE_OFFSET);
+
+            if (
+                logKey >= uint256(SystemLogKey.BLOB_ONE_HASH_KEY) &&
+                logKey != uint256(SystemLogKey.EXPECTED_SYSTEM_CONTRACT_UPGRADE_TX_HASH_KEY)
+            ) {
+                uint8 key = uint8(logKey) - uint8(SystemLogKey.BLOB_ONE_HASH_KEY);
+                // Ensure that the log hasn't been processed already
+                require(!_checkBit(processedLogs, key), "pk");
+                processedLogs = _setBit(processedLogs, key);
+
+                require(logSender == L2_PUBDATA_CHUNK_PUBLISHER_ADDR, "pc");
+                blobHashes[logKey - uint256(SystemLogKey.BLOB_ONE_HASH_KEY)] = logValue;
```

Comment:
could you please replicate the changes made in [this](https://github.com/matter-labs/era-contracts/pull/251) PR after @vladbochok comments?